ethereum/contracts/pyth/PythGovernance.sol

```diff
@@ -67,13 +70,33 @@ abstract contract PythGovernance is PythGetters, PythSetters, PythGovernanceInst
 
     function upgradeUpgradableContract(UpgradeContractPayload memory payload) virtual internal;
 
-    function setGovernanceDataSource(bytes memory encodedPayload) internal {
-        SetGovernanceDataSourcePayload memory payload = parseSetGovernanceDataSourcePayload(encodedPayload);
-
+    // Transfer the governance data source to a new value with sanity checks
+    // to ensure the new governance data source can manage the contract.
+    function transferGovernanceDataSource(bytes memory encodedPayload) internal {
         PythInternalStructs.DataSource memory oldGovernanceDatSource = governanceDataSource();
 
-        setGovernanceDataSource(payload.newGovernanceDataSource);
-        setLastExecutedGovernanceSequence(payload.initialSequence);
+        TransferGovernanceDataSourcePayload memory payload = parseTransferGovernanceDataSourcePayload(encodedPayload);
+
+        // Make sure the VAA is valid
+        (IWormhole.VM memory vm, bool valid, string memory reason) = wormhole().parseAndVerifyVM(payload.claimVaa);
+        require(valid, reason);
+
+        GovernanceInstruction memory gi = parseGovernanceInstruction(vm.payload);
+        require(gi.targetChainId == 0, "governance data source change claim vaa should be applied to all chains");
+        require(gi.action == GovernanceAction.TransferGovernanceDataSourceClaim,
+            "governance data source change inner vaa is not of claim action type");
+
+        TransferGovernanceDataSourceClaimPayload memory claimPayload = parseTransferGovernanceDataSourceClaimPayload(gi.payload);
+
+        require(governanceDataSourceIndex() < claimPayload.governanceDataSourceIndex, 
+            "cannot upgrade to an older governance data source");
+
+        setGovernanceDataSourceIndex(claimPayload.governanceDataSourceIndex);
```

Comment:
You are not missing anything here. As I said in the PR description its an extra sanity check. It's not preventing any obvious replay attacks. We want to reduce the chance of mistake anywhere possible.

Based on my understanding the goal of governance is the be restricted as much as possible. It is better to have an index to be able to point to the version. I believe that we are making it very difficult and I personally don't like it, but I think this is the security decision. 

Now this is the case that a replay attack it can happen which is non-obvious. Imagine we go from A to B and from B to A again.
Then after that one can use the old claim of B (your naming: approval) and wrap it into a new transfer of A (your naming: request) and do transfer the governance to B again, but with the old sequence coming as the result.